diff -ruN a/nordicsemi/__main__.py b/nordicsemi/__main__.py
--- a/nordicsemi/__main__.py
+++ b/nordicsemi/__main__.py
@@ -58,7 +58,6 @@
 from nordicsemi.zigbee.prod_config import ProductionConfig, ProductionConfigWrongException, ProductionConfigTooLargeException
 from pc_ble_driver_py.exceptions import NordicSemiException, NotImplementedException
 from nordicsemi.lister.device_lister import DeviceLister
-import spinel.util as util
 
 
 logger = logging.getLogger(__name__)
@@ -1364,6 +1363,7 @@
     if (flash_connectivity):
         config[tncp.NCPTransport.CFG_KEY_RESET] = False
     if (masterkey):
+        import spinel.util as util
         config[tncp.NCPTransport.CFG_KEY_MASTERKEY] = util.hex_to_bytes(masterkey)
 
     opts = type('DFUServerOptions', (object,), {})()
diff -ruN a/setup.py b/setup.py
--- a/setup.py
+++ b/setup.py
@@ -47,7 +47,6 @@
 import sys
 
 from setuptools import setup, find_packages
-from setuptools.command.test import test as TestCommand
 
 from nordicsemi import version
 
@@ -118,16 +117,6 @@
 with open("requirements.txt") as reqs_file:
     reqs = reqs_file.readlines()
 
-class NoseTestCommand(TestCommand):
-    def finalize_options(self):
-        TestCommand.finalize_options(self)
-        self.test_args = []
-        self.test_suite = True
-
-    def run_tests(self):
-        import nose
-        nose.run_exit(argv=['nosetests', '--with-xunit', '--xunit-file=test-reports/unittests.xml'])
-
 setup(
     name="nrfutil",
     version=version.NRFUTIL_VERSION,
@@ -167,9 +156,6 @@
         'Programming Language :: Python :: 2.7',
     ],
     keywords = 'nordic nrf51 nrf52 ble bluetooth dfu ota softdevice serialization nrfutil pc-nrfutil',
-    cmdclass={
-        'test': NoseTestCommand
-    },
     entry_points='''
       [console_scripts]
       nrfutil = nordicsemi.__main__:cli
